File: android/src/main/kotlin/br/com/cpndntech/google_maps_plus/Convert.kt
package br.com.cpndntech.google_maps_plus

import android.graphics.Color
import com.google.android.gms.maps.GoogleMap
import com.google.android.gms.maps.model.*

// --- Classes de Dados ---

data class GoogleMapsPlaybackPoint(
    val lat: Double,
    val lng: Double,
    val bearing: Double,
    val isStop: Boolean
)

data class MapSettings(
    val mapType: Int = GoogleMap.MAP_TYPE_NORMAL,
    val showTraffic: Boolean = false,
    val showBuildings: Boolean = true,
    val showUserLocation: Boolean = false,
    val showMyLocationButton: Boolean = true,
    val compassEnabled: Boolean = true,
    val mapToolbarEnabled: Boolean = true,
    val rotateGesturesEnabled: Boolean = true,
    val scrollGesturesEnabled: Boolean = true,
    val zoomControlsEnabled: Boolean = true,
    val zoomGesturesEnabled: Boolean = true,
    val tiltGesturesEnabled: Boolean = true,
    val indoorViewEnabled: Boolean = false,
    val isDark: Boolean = false,
    val style: String? = null,
    val padding: List<Double>? = null,
    val initialMarkers: List<Map<String, Any>>? = null,
    val initialCircles: List<Map<String, Any>>? = null,
    val initialPolylines: List<Map<String, Any>>? = null,
    val initialPolygons: List<Map<String, Any>>? = null,
    val defaultSpeed: Double = 60.0,
    val maxAnimationDuration: Long = 2000L
)

data class PlaybackSettings(
    val baseSpeed: Double = 60.0,
    val canRotate: Boolean = true,
    val dynamicRotation: Boolean = false,
    val showStops: Boolean = false,
    val vehicleIcon: Any? = null,
    val stopIcon: Any? = null,
    val drawTrail: Boolean = true,
    val polylineColor: Any? = null,
    val points: List<Map<String, Any>>? = null,
    val autoStart: Boolean = false
)

// --- Utilitários de Conversão ---

object Convert {
    fun toDouble(o: Any?): Double? {
        return (o as? Number)?.toDouble()
    }

    fun toMapSettings(data: Any?): MapSettings {
        val m = data as? Map<String, Any> ?: return MapSettings()
        return MapSettings(
            mapType = m.getInt("mapType", GoogleMap.MAP_TYPE_NORMAL),
            showTraffic = m.getBool("showTraffic", false),
            showBuildings = m.getBool("showBuildings", true),
            showUserLocation = m.getBool("showUserLocation", false),
            showMyLocationButton = m.getBool("showMyLocationButton", true),
            compassEnabled = m.getBool("compassEnabled", true),
            mapToolbarEnabled = m.getBool("mapToolbarEnabled", true),
            rotateGesturesEnabled = m.getBool("rotateGesturesEnabled", true),
            scrollGesturesEnabled = m.getBool("scrollGesturesEnabled", true),
            zoomControlsEnabled = m.getBool("zoomControlsEnabled", true),
            zoomGesturesEnabled = m.getBool("zoomGesturesEnabled", true),
            tiltGesturesEnabled = m.getBool("tiltGesturesEnabled", true),
            indoorViewEnabled = m.getBool("indoorViewEnabled", false),
            isDark = m.getBool("isDark", false),
            style = m.getString("style"),
            padding = m.getList("padding"),
            initialMarkers = m.getList("markers"),
            initialCircles = m.getList("circles"),
            initialPolylines = m.getList("polylines"),
            initialPolygons = m.getList("polygons"),
            defaultSpeed = m.getDouble("defaultSpeed", 60.0),
            maxAnimationDuration = m.getLong("maxAnimationDuration", 2000L)
        )
    }

    fun toPlaybackSettings(data: Any?): PlaybackSettings {
        val m = data as? Map<String, Any> ?: return PlaybackSettings()
        return PlaybackSettings(
            baseSpeed = m.getDouble("baseSpeed", 60.0),
            canRotate = m.getBool("canRotate", true),
            dynamicRotation = m.getBool("dynamicRotation", false),
            showStops = m.getBool("showStops", false),
            vehicleIcon = m["vehicleIcon"],
            stopIcon = m["stopIcon"],
            drawTrail = m.getBool("drawTrail", true),
            polylineColor = m["polylineColor"],
            points = m.getList("points"),
            autoStart = m.getBool("autoStart", false)
        )
    }

    // --- Helpers de Extensão para Map ---

    private fun Map<String, Any>.getInt(key: String, default: Int): Int = (this[key] as? Number)?.toInt() ?: default
    private fun Map<String, Any>.getBool(key: String, default: Boolean): Boolean = this[key] as? Boolean ?: default
    private fun Map<String, Any>.getDouble(key: String, default: Double): Double = (this[key] as? Number)?.toDouble() ?: default
    private fun Map<String, Any>.getLong(key: String, default: Long): Long = (this[key] as? Number)?.toLong() ?: default
    private fun Map<String, Any>.getString(key: String): String? = this[key] as? String
    private fun <T> Map<String, Any>.getList(key: String): List<T>? = this[key] as? List<T>

    // --- Outras Conversões (Marcadores, etc) ---

    fun toMarkerUpdates(data: Any?): MapObjectUpdates<MarkerData> {
        val map = data as? Map<String, Any> ?: return MapObjectUpdates(emptyList(), emptyList(), emptyList())
        val add = (map["toAdd"] as? List<Map<String, Any>>)?.map { MarkerData(it["markerId"] as String, it) } ?: emptyList()
        val change = (map["toChange"] as? List<Map<String, Any>>)?.map { MarkerData(it["markerId"] as String, it) } ?: emptyList()
        val remove = (map["toRemove"] as? List<String>) ?: emptyList()
        return MapObjectUpdates(add, change, remove)
    }

    fun toPolylineUpdates(data: Any?): MapObjectUpdates<PolylineData> {
        val map = data as? Map<String, Any> ?: return MapObjectUpdates(emptyList(), emptyList(), emptyList())
        val add = (map["toAdd"] as? List<Map<String, Any>>)?.map { PolylineData(it["polylineId"] as String, it) } ?: emptyList()
        val change = (map["toChange"] as? List<Map<String, Any>>)?.map { PolylineData(it["polylineId"] as String, it) } ?: emptyList()
        val remove = (map["toRemove"] as? List<String>) ?: emptyList()
        return MapObjectUpdates(add, change, remove)
    }

    fun toCircleUpdates(data: Any?): MapObjectUpdates<CircleData> {
        val map = data as? Map<String, Any> ?: return MapObjectUpdates(emptyList(), emptyList(), emptyList())
        val add = (map["toAdd"] as? List<Map<String, Any>>)?.map { CircleData(it["circleId"] as String, it) } ?: emptyList()
        val change = (map["toChange"] as? List<Map<String, Any>>)?.map { CircleData(it["circleId"] as String, it) } ?: emptyList()
        val remove = (map["toRemove"] as? List<String>) ?: emptyList()
        return MapObjectUpdates(add, change, remove)
    }

    fun toPolygonUpdates(data: Any?): MapObjectUpdates<PolygonData> {
        val map = data as? Map<String, Any> ?: return MapObjectUpdates(emptyList(), emptyList(), emptyList())
        val add = (map["toAdd"] as? List<Map<String, Any>>)?.map { PolygonData(it["polygonId"] as String, it) } ?: emptyList()
        val change = (map["toChange"] as? List<Map<String, Any>>)?.map { PolygonData(it["polygonId"] as String, it) } ?: emptyList()
        val remove = (map["toRemove"] as? List<String>) ?: emptyList()
        return MapObjectUpdates(add, change, remove)
    }

    fun toLatLng(data: Any?): LatLng? {
        if (data is List<*>) {
            if (data.size < 2) return null
            val lat = (data[0] as? Number)?.toDouble() ?: return null
            val lng = (data[1] as? Number)?.toDouble() ?: return null
            return LatLng(lat, lng)
        }
        if (data is Map<*, *>) {
            val lat = (data["latitude"] ?: data["lat"]) as? Number ?: return null
            val lng = (data["longitude"] ?: data["lng"]) as? Number ?: return null
            return LatLng(lat.toDouble(), lng.toDouble())
        }
        return null
    }

    fun toColor(data: Any?): Int {
        if (data == null) return Color.BLUE
        if (data is String) return try { Color.parseColor(data) } catch (e: Exception) { Color.BLUE }
        if (data is Number) return data.toInt()
        return Color.BLUE
    }

    fun toBitmapDescriptor(data: Any?): BitmapDescriptor? {
        val list = data as? List<*> ?: return null
        if (list.isEmpty()) return null
        val type = list[0] as? String ?: return null
        return when (type) {
            "defaultMarker" -> {
                val hue = if (list.size > 1) (list[1] as? Number)?.toFloat() ?: 0.0f else 0.0f
                BitmapDescriptorFactory.defaultMarker(hue)
            }
            "fromAsset", "fromAssetImage" -> BitmapDescriptorFactory.fromAsset(list[1] as? String ?: return null)
            "fromBytes" -> {
                val bytes = list[1] as? ByteArray ?: return null
                BitmapDescriptorFactory.fromBitmap(android.graphics.BitmapFactory.decodeByteArray(bytes, 0, bytes.size))
            }
            "asset" -> BitmapDescriptorFactory.fromAsset((list[1] as? Map<*, *>)?.get("assetName") as? String ?: return null)
            "bytes" -> {
                val bytes = (list[1] as? Map<*, *>)?.get("byteData") as? ByteArray ?: return null
                BitmapDescriptorFactory.fromBitmap(android.graphics.BitmapFactory.decodeByteArray(bytes, 0, bytes.size))
            }
            else -> null
        }
    }

    fun toMarkerOptions(data: Map<String, Any>): MarkerOptions {
        val options = MarkerOptions()
        toLatLng(data["position"])?.let { options.position(it) }
        (data["anchor"] as? List<*>)?.let { options.anchor((it[0] as? Number)?.toFloat() ?: 0.5f, (it[1] as? Number)?.toFloat() ?: 0.5f) }
        (data["rotation"] as? Number)?.let { options.rotation(it.toFloat()) }
        (data["zIndex"] as? Number)?.let { options.zIndex(it.toFloat()) }
        (data["flat"] as? Boolean)?.let { options.flat(it) }
        (data["alpha"] as? Number)?.let { options.alpha(it.toFloat()) }
        toBitmapDescriptor(data["icon"])?.let { options.icon(it) }
        return options
    }

    fun toCircleOptions(data: Map<String, Any>, density: Float): CircleOptions {
        val options = CircleOptions()
        toLatLng(data["center"])?.let { options.center(it) }
        (data["radius"] as? Number)?.let { options.radius(it.toDouble()) }
        options.strokeWidth(((data["strokeWidth"] as? Number)?.toFloat() ?: 10f) * density)
        options.strokeColor(toColor(data["strokeColor"]))
        options.fillColor(toColor(data["fillColor"]))
        options.zIndex((data["zIndex"] as? Number)?.toFloat() ?: 0f)
        options.visible(data["visible"] as? Boolean ?: true)
        return options
    }

    fun toPolylineOptions(data: Map<String, Any>, density: Float): PolylineOptions {
        val options = PolylineOptions()
        (data["points"] as? List<*>)?.mapNotNull { toLatLng(it) }?.let { options.addAll(it) }
        options.width(((data["width"] as? Number)?.toFloat() ?: 10f) * density)
        options.color(toColor(data["color"]))
        options.zIndex((data["zIndex"] as? Number)?.toFloat() ?: 0f)
        options.geodesic(data["geodesic"] as? Boolean ?: false)
        options.visible(data["visible"] as? Boolean ?: true)
        return options
    }

    fun toPolygonOptions(data: Map<String, Any>, density: Float): PolygonOptions {
        val options = PolygonOptions()
        (data["points"] as? List<*>)?.mapNotNull { toLatLng(it) }?.let { options.addAll(it) }
        options.strokeWidth(((data["strokeWidth"] as? Number)?.toFloat() ?: 10f) * density)
        options.strokeColor(toColor(data["strokeColor"]))
        options.fillColor(toColor(data["fillColor"]))
        options.zIndex((data["zIndex"] as? Number)?.toFloat() ?: 0f)
        options.geodesic(data["geodesic"] as? Boolean ?: false)
        options.visible(data["visible"] as? Boolean ?: true)
        return options
    }

    fun interpretMarker(m: Map<String, Any>, marker: Marker) {
        toLatLng(m["position"])?.let { marker.position = it }
        (m["anchor"] as? List<*>)?.let { marker.setAnchor((it[0] as? Number)?.toFloat() ?: 0.5f, (it[1] as? Number)?.toFloat() ?: 0.5f) }
        (m["rotation"] as? Number)?.let { marker.rotation = it.toFloat() }
        (m["zIndex"] as? Number)?.let { marker.zIndex = it.toFloat() }
        (m["flat"] as? Boolean)?.let { marker.isFlat = it }
        (m["alpha"] as? Number)?.let { marker.alpha = it.toFloat() }
    }

    fun interpretCircle(c: Map<String, Any>, circle: Circle, density: Float) {
        toLatLng(c["center"])?.let { circle.center = it }
        (c["radius"] as? Number)?.let { circle.radius = it.toDouble() }
        (c["strokeWidth"] as? Number)?.let { circle.strokeWidth = it.toFloat() * density }
        circle.strokeColor = toColor(c["strokeColor"])
        circle.fillColor = toColor(c["fillColor"])
        (c["zIndex"] as? Number)?.let { circle.zIndex = it.toFloat() }
        (c["visible"] as? Boolean)?.let { circle.isVisible = it }
    }

    fun interpretPolyline(p: Map<String, Any>, polyline: Polyline, density: Float) {
        (p["points"] as? List<*>)?.mapNotNull { toLatLng(it) }?.let { polyline.points = it }
        (p["width"] as? Number)?.let { polyline.width = it.toFloat() * density }
        polyline.color = toColor(p["color"])
        (p["zIndex"] as? Number)?.let { polyline.zIndex = it.toFloat() }
        (p["geodesic"] as? Boolean)?.let { polyline.isGeodesic = it }
        (p["visible"] as? Boolean)?.let { polyline.isVisible = it }
    }

    fun interpretPolygon(p: Map<String, Any>, polygon: Polygon, density: Float) {
        (p["points"] as? List<*>)?.mapNotNull { toLatLng(it) }?.let { polygon.points = it }
        (p["strokeWidth"] as? Number)?.let { polygon.strokeWidth = it.toFloat() * density }
        polygon.strokeColor = toColor(p["strokeColor"])
        polygon.fillColor = toColor(p["fillColor"])
        (p["zIndex"] as? Number)?.let { polygon.zIndex = it.toFloat() }
        (p["geodesic"] as? Boolean)?.let { polygon.isGeodesic = it }
        (p["visible"] as? Boolean)?.let { polygon.isVisible = it }
    }
}

// --- Estruturas Auxiliares ---

data class MapObjectUpdates<T>(val toAdd: List<T>, val toChange: List<T>, val toRemove: List<String>)
data class MarkerData(val id: String, val data: Map<String, Any>)
data class PolylineData(val id: String, val data: Map<String, Any>)
data class CircleData(val id: String, val data: Map<String, Any>)
data class PolygonData(val id: String, val data: Map<String, Any>)

object MethodNames {
    const val UPDATE_OPTIONS = "updateOptions"
    const val MARKERS_UPDATE = "markers/update"
    const val POLYLINES_UPDATE = "polylines/update"
    const val CIRCLES_UPDATE = "circles/update"
    const val POLYGONS_UPDATE = "polygons/update"
    const val MARKER_ADD = "marker_add"
    const val MARKER_MOVE = "marker_move"
    const val MARKER_ICON = "marker_icon"
    const val MARKER_REMOVE = "marker_remove"
    const val MARKER_SHOW_INFO_WINDOW = "marker_show_info_window"
    const val CIRCLE_ADD = "circle_add"
    const val CIRCLE_REMOVE = "circle_remove"
    const val POLYLINE_ADD = "polyline_add"
    const val POLYLINE_REMOVE = "polyline_remove"
    const val POLYGON_ADD = "polygon_add"
    const val POLYGON_REMOVE = "polygon_remove"
    const val ZOOM_IN = "zoomIn"
    const val ZOOM_OUT = "zoomOut"
    const val FOLLOW_MARKER = "follow_marker"
    const val MOVE_CAMERA = "move_camera"
    const val SET_DEFAULT_SPEED = "set_default_speed"
    const val SET_MAX_ANIMATION_DURATION = "set_max_animation_duration"
}

File: ios/Classes/Convert.swift
import Foundation
import GoogleMaps
import Flutter

// --- Modelos de Dados ---

struct GoogleMapsPlaybackPoint {
    let lat: Double
    let lng: Double
    let bearing: Double
    let isStop: Bool
}

struct MapSettings {
    let mapType: Int
    let showTraffic: Bool
    let showBuildings: Bool
    let showUserLocation: Bool
    let showMyLocationButton: Bool
    let compassEnabled: Bool
    let rotateGesturesEnabled: Bool
    let scrollGesturesEnabled: Bool
    let zoomGesturesEnabled: Bool
    let tiltGesturesEnabled: Bool
    let indoorViewEnabled: Bool
    let isDark: Bool
    let style: String?
    let padding: [Double]?
    let initialMarkers: [[String: Any]]?
    let initialCircles: [[String: Any]]?
    let initialPolylines: [[String: Any]]?
    let initialPolygons: [[String: Any]]?
    let defaultSpeed: Double
    let maxAnimationDuration: Double
}

struct PlaybackSettings {
    let baseSpeed: Double
    let canRotate: Bool
    let dynamicRotation: Bool
    let showStops: Bool
    let vehicleIcon: Any?
    let stopIcon: Any?
    let drawTrail: Bool
    let polylineColor: Any?
    let points: [[String: Any]]?
    let autoStart: Bool
}

// --- Conversor ---

class Convert {
    
    static func toMapSettings(_ data: Any?) -> MapSettings {
        let m = data as? [String: Any] ?? [:]
        return MapSettings(
            mapType: m.getInt("mapType", 1),
            showTraffic: m.getBool("showTraffic", false),
            showBuildings: m.getBool("showBuildings", true),
            showUserLocation: m.getBool("showUserLocation", false),
            showMyLocationButton: m.getBool("showMyLocationButton", true),
            compassEnabled: m.getBool("compassEnabled", true),
            rotateGesturesEnabled: m.getBool("rotateGesturesEnabled", true),
            scrollGesturesEnabled: m.getBool("scrollGesturesEnabled", true),
            zoomGesturesEnabled: m.getBool("zoomGesturesEnabled", true),
            tiltGesturesEnabled: m.getBool("tiltGesturesEnabled", true),
            indoorViewEnabled: m.getBool("indoorViewEnabled", false),
            isDark: m.getBool("isDark", false),
            style: m.getString("style"),
            padding: m.getList("padding"),
            initialMarkers: m.getList("markers"),
            initialCircles: m.getList("circles"),
            initialPolylines: m.getList("polylines"),
            initialPolygons: m.getList("polygons"),
            defaultSpeed: m.getDouble("defaultSpeed", 60.0),
            maxAnimationDuration: m.getDouble("maxAnimationDuration", 2000.0) / 1000.0
        )
    }

    static func toPlaybackSettings(_ data: Any?) -> PlaybackSettings {
        let m = data as? [String: Any] ?? [:]
        return PlaybackSettings(
            baseSpeed: m.getDouble("baseSpeed", 60.0),
            canRotate: m.getBool("canRotate", true),
            dynamicRotation: m.getBool("dynamicRotation", false),
            showStops: m.getBool("showStops", false),
            vehicleIcon: m["vehicleIcon"],
            stopIcon: m["stopIcon"],
            drawTrail: m.getBool("drawTrail", true),
            polylineColor: m["polylineColor"],
            points: m.getList("points"),
            autoStart: m.getBool("autoStart", false)
        )
    }

    // --- Outras Conversões ---

    static func toMarkerUpdates(_ data: Any?) -> MapObjectUpdates<MarkerData> {
        guard let map = data as? [String: Any] else { return MapObjectUpdates(toAdd: [], toChange: [], toRemove: []) }
        let add = (map["toAdd"] as? [[String: Any]])?.map { MarkerData(id: $0["markerId"] as! String, data: $0) } ?? []
        let change = (map["toChange"] as? [[String: Any]])?.map { MarkerData(id: $0["markerId"] as! String, data: $0) } ?? []
        let remove = (map["toRemove"] as? [String]) ?? []
        return MapObjectUpdates(toAdd: add, toChange: change, toRemove: remove)
    }

    static func toPolylineUpdates(_ data: Any?) -> MapObjectUpdates<PolylineData> {
        guard let map = data as? [String: Any] else { return MapObjectUpdates(toAdd: [], toChange: [], toRemove: []) }
        let add = (map["toAdd"] as? [[String: Any]])?.map { PolylineData(id: $0["polylineId"] as! String, data: $0) } ?? []
        let change = (map["toChange"] as? [[String: Any]])?.map { PolylineData(id: $0["polylineId"] as! String, data: $0) } ?? []
        let remove = (map["toRemove"] as? [String]) ?? []
        return MapObjectUpdates(toAdd: add, toChange: change, toRemove: remove)
    }

    static func toCircleUpdates(_ data: Any?) -> MapObjectUpdates<CircleData> {
        guard let map = data as? [String: Any] else { return MapObjectUpdates(toAdd: [], toChange: [], toRemove: []) }
        let add = (map["toAdd"] as? [[String: Any]])?.map { CircleData(id: $0["circleId"] as! String, data: $0) } ?? []
        let change = (map["toChange"] as? [[String: Any]])?.map { CircleData(id: $0["circleId"] as! String, data: $0) } ?? []
        let remove = (map["toRemove"] as? [String]) ?? []
        return MapObjectUpdates(toAdd: add, toChange: change, toRemove: remove)
    }

    static func toPolygonUpdates(_ data: Any?) -> MapObjectUpdates<PolygonData> {
        guard let map = data as? [String: Any] else { return MapObjectUpdates(toAdd: [], toChange: [], toRemove: []) }
        let add = (map["toAdd"] as? [[String: Any]])?.map { PolygonData(id: $0["polygonId"] as! String, data: $0) } ?? []
        let change = (map["toChange"] as? [[String: Any]])?.map { PolygonData(id: $0["polygonId"] as! String, data: $0) } ?? []
        let remove = (map["toRemove"] as? [String]) ?? []
        return MapObjectUpdates(toAdd: add, toChange: change, toRemove: remove)
    }

    static func toCoordinate(_ data: Any?) -> CLLocationCoordinate2D? {
        guard let list = data as? [Any], list.count >= 2 else { return nil }
        guard let lat = toDouble(list[0]), let lng = toDouble(list[1]) else { return nil }
        return CLLocationCoordinate2D(latitude: lat, longitude: lng)
    }

    static func toDouble(_ data: Any?) -> Double? {
        if let d = data as? Double { return d }
        if let ns = data as? NSNumber { return ns.doubleValue }
        return nil
    }
    
    static func toColor(_ data: Any?) -> UIColor {
        let colorInt: Int64?
        if let v = data as? Int64 { colorInt = v }
        else if let v = data as? Int { colorInt = Int64(v) }
        else { return .blue }
        let c = colorInt!
        return UIColor(red: CGFloat((c >> 16) & 0xFF) / 255.0, green: CGFloat((c >> 8) & 0xFF) / 255.0, blue: CGFloat(c & 0xFF) / 255.0, alpha: CGFloat((c >> 24) & 0xFF) / 255.0)
    }
    
    static func toIcon(_ data: Any?, registrar: FlutterPluginRegistrar? = nil) -> UIImage? {
        guard let list = data as? [Any], !list.isEmpty else { return nil }
        let type = list[0] as? String
        let screenScale = UIScreen.main.scale
        switch type {
        case "fromAsset":
            if let assetName = list[1] as? String {
                return loadAssetImage(assetName, registrar: registrar)
            }
        case "fromAssetImage":
            // list: [type, assetName, pixelRatio, size:[w,h]]
            if let assetName = list[1] as? String {
                let image = loadAssetImage(assetName, registrar: registrar)
                // Respeita o size explícito em pontos lógicos, ignora pixelRatio
                if let img = image, list.count >= 4,
                   let sizeList = list[3] as? [Any], sizeList.count >= 2,
                   let w = toDouble(sizeList[0]), let h = toDouble(sizeList[1]), w > 0, h > 0 {
                    return resizeImage(img, to: CGSize(width: w, height: h))
                }
                return image
            }
        case "fromBytes":
            // Bytes já estão na resolução física da tela; aplicar screenScale para obter tamanho lógico correto
            if let bytes = list[1] as? FlutterStandardTypedData {
                if let image = UIImage(data: bytes.data, scale: screenScale) {
                    return fixOrientation(image)
                }
            }
        case "asset":
            if let params = list[1] as? [String: Any], let assetName = params["assetName"] as? String {
                return loadAssetImage(assetName, registrar: registrar)
            }
        case "bytes":
            if let params = list[1] as? [String: Any], let bytes = params["byteData"] as? FlutterStandardTypedData {
                if let image = UIImage(data: bytes.data, scale: screenScale) {
                    return fixOrientation(image)
                }
            }
        default: return nil
        }
        return nil
    }

    private static func loadAssetImage(_ assetName: String, registrar: FlutterPluginRegistrar?) -> UIImage? {
        if let reg = registrar {
            let key = reg.lookupKey(forAsset: assetName)
            if let path = Bundle.main.path(forResource: key, ofType: nil) {
                return UIImage(contentsOfFile: path)
            }
        }
        return UIImage(named: assetName)
    }

    private static func fixOrientation(_ image: UIImage) -> UIImage {
        guard image.imageOrientation != .up else { return image }
        UIGraphicsBeginImageContextWithOptions(image.size, false, image.scale)
        image.draw(in: CGRect(origin: .zero, size: image.size))
        let result = UIGraphicsGetImageFromCurrentImageContext()
        UIGraphicsEndImageContext()
        return result ?? image
    }

    private static func resizeImage(_ image: UIImage, to size: CGSize) -> UIImage {
        UIGraphicsBeginImageContextWithOptions(size, false, UIScreen.main.scale)
        image.draw(in: CGRect(origin: .zero, size: size))
        let result = UIGraphicsGetImageFromCurrentImageContext()
        UIGraphicsEndImageContext()
        return result ?? image
    }
    
    static func interpretMarker(_ m: [String: Any], marker: GMSMarker, registrar: FlutterPluginRegistrar? = nil) {
        if let coord = toCoordinate(m["position"]) { marker.position = coord }
        if let anchor = m["anchor"] as? [Any], anchor.count >= 2 { marker.groundAnchor = CGPoint(x: toDouble(anchor[0]) ?? 0.5, y: toDouble(anchor[1]) ?? 0.5) }
        if let rotation = toDouble(m["rotation"]) { marker.rotation = rotation }
        if let zIndex = toDouble(m["zIndex"]) { marker.zIndex = Int32(zIndex) }
        if let flat = m["flat"] as? Bool { marker.isFlat = flat }
        if let opacity = toDouble(m["alpha"]) { marker.opacity = Float(opacity) }
        if let iconData = m["icon"] { marker.icon = toIcon(iconData, registrar: registrar) }
    }
    
    static func interpretCircle(_ c: [String: Any], circle: GMSCircle) {
        if let coord = toCoordinate(c["center"]) { circle.position = coord }
        if let radius = toDouble(c["radius"]) { circle.radius = radius }
        circle.fillColor = toColor(c["fillColor"]); circle.strokeColor = toColor(c["strokeColor"])
        if let strokeWidth = toDouble(c["strokeWidth"]) { circle.strokeWidth = CGFloat(strokeWidth) }
        if let zIndex = toDouble(c["zIndex"]) { circle.zIndex = Int32(zIndex) }
    }
    
    static func interpretPolyline(_ p: [String: Any], polyline: GMSPolyline) {
        if let pts = p["points"] as? [Any] {
            let path = GMSMutablePath()
            for pt in pts { if let coord = toCoordinate(pt) { path.add(coord) } }
            polyline.path = path
        }
        polyline.strokeColor = toColor(p["color"])
        if let width = toDouble(p["width"]) { polyline.strokeWidth = CGFloat(width) }
        if let zIndex = toDouble(p["zIndex"]) { polyline.zIndex = Int32(zIndex) }
        if let geodesic = p["geodesic"] as? Bool { polyline.geodesic = geodesic }
    }
    
    static func interpretPolygon(_ poly: [String: Any], polygon: GMSPolygon) {
        if let pts = poly["points"] as? [Any] {
            let path = GMSMutablePath()
            for pt in pts { if let coord = toCoordinate(pt) { path.add(coord) } }
            polygon.path = path
        }
        polygon.strokeColor = toColor(poly["strokeColor"]); polygon.fillColor = toColor(poly["fillColor"])
        if let width = toDouble(poly["strokeWidth"]) { polygon.strokeWidth = CGFloat(width) }
        if let zIndex = toDouble(poly["zIndex"]) { polygon.zIndex = Int32(zIndex) }
        if let geodesic = poly["geodesic"] as? Bool { polygon.geodesic = geodesic }
    }
}

// --- Helpers de Dicionário ---

extension Dictionary where Key == String {
    func getInt(_ key: String, _ defaultVal: Int) -> Int { return self[key] as? Int ?? defaultVal }
    func getBool(_ key: String, _ defaultVal: Bool) -> Bool { return self[key] as? Bool ?? defaultVal }
    func getDouble(_ key: String, _ defaultVal: Double) -> Double { return Convert.toDouble(self[key]) ?? defaultVal }
    func getString(_ key: String) -> String? { return self[key] as? String }
    func getList<T>(_ key: String) -> [T]? { return self[key] as? [T] }
}

// --- Estruturas Auxiliares ---

struct MapObjectUpdates<T> { let toAdd: [T]; let toChange: [T]; let toRemove: [String] }
struct MarkerData { let id: String; let data: [String: Any] }
struct PolylineData { let id: String; let data: [String: Any] }
struct CircleData { let id: String; let data: [String: Any] }
struct PolygonData { let id: String; let data: [String: Any] }

struct MethodNames {
    static let updateOptions = "updateOptions"
    static let markersUpdate = "markers/update"
    static let polylinesUpdate = "polylines/update"
    static let circlesUpdate = "circles/update"
    static let polygonsUpdate = "polygons/update"
    static let markerAdd = "marker_add"; static let markerMove = "marker_move"; static let markerIcon = "marker_icon"; static let markerRemove = "marker_remove"; static let markerShowInfoWindow = "marker_show_info_window"
    static let circleAdd = "circle_add"; static let circleRemove = "circle_remove"
    static let polylineAdd = "polyline_add"; static let polylineRemove = "polyline_remove"
    static let polygonAdd = "polygon_add"; static let polygonRemove = "polygon_remove"
    static let zoomIn = "zoomIn"; static let zoomOut = "zoomOut"; static let followMarker = "follow_marker"; static let moveCamera = "move_camera"
    static let setDefaultSpeed = "set_default_speed"; static let setMaxAnimationDuration = "set_max_animation_duration"
}
